packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybrid2GL.hpp

```diff
@@ -0,0 +1,1532 @@
+				 typename Kokkos::View<lno_t*,device_type>::HostMirror verts_to_send_atomic,
+				 typename Kokkos::View<size_t*,device_type>::HostMirror verts_to_send_size_atomic,
+				 typename Kokkos::View<gno_t*, device_type>::HostMirror recoloringSize,
+			         typename Kokkos::View<int*,  device_type>::HostMirror rand,
+			         typename Kokkos::View<gno_t*,device_type>::HostMirror gid,
+                                 typename Kokkos::View<gno_t*,device_type>::HostMirror ghost_degrees,
+				 bool recolor_degrees) = 0;
+    //Entry point for the construction of the boundary
+    //INPUT ARGS
+    //  n_local is the number of vertices owned by the current process
+    //
+    //  dist_offsets_dev is the device view that holds the CSR offsets
+    //
+    //  dist_adjs_dev is the device view that holds CSR adjacencies
+    //
+    //  dist_offsets_host is the hostmirror that holds the CSR offsets
+    //   
+    //  dist_adjs_host is the hostmirror that holds the CSR adjacencies
+    //
+    //OUTPUT ARGS
+    //  boundary_verts is an unallocated device view that will hold the
+    //                 list of boundary vertices.
+    //
+    //  boundary_size will hold the size of the list of boundary vertices
+    //
+    //  verts_to_send_atomic will hold the list of vertices to send
+    //
+    //  verts_to_send_size_atomic will hold the size of the list of
+    //                            vertices to send.
+    //
+    virtual void constructBoundary(const size_t n_local,
+		    		   Kokkos::View<offset_t*, device_type> dist_offsets_dev,
+		                   Kokkos::View<lno_t*, device_type> dist_adjs_dev,
+				   typename Kokkos::View<offset_t*, device_type>::HostMirror dist_offsets_host,
+				   typename Kokkos::View<lno_t*, device_type>::HostMirror dist_adjs_host,
+                                   Kokkos::View<lno_t*, device_type>& boundary_verts,
+				   gno_t& boundary_size,
+				   Kokkos::View<lno_t*, 
+				                device_type, 
+						Kokkos::MemoryTraits<Kokkos::Atomic> > verts_to_send_atomic,
+				   Kokkos::View<size_t*, 
+				                device_type, 
+						Kokkos::MemoryTraits<Kokkos::Atomic>> verts_to_send_size_atomic) = 0;
+				   
+  protected:
+    RCP<const base_adapter_t> adapter;
+    RCP<GraphModel<base_adapter_t> > model;
+    RCP<Teuchos::ParameterList> pl;
+    RCP<Environment> env;
+    RCP<const Teuchos::Comm<int> > comm;
+    bool verbose;
+
+    
+  private:
+    //This function constructs the second ghost layer.
+    //It proceeds in rounds in order to keep messages
+    //under MPI's maximum send size.
+    //
+    //OUTPUT ARGS
+    //  ownedPlusGhosts holds the initial list of GIDs for all owned vertices.
+    //                  This function changes the order of the ghost GIDs in
+    //                  order to more easily construct the second ghost layer.
+    //
+    //  adjs_2GL holds the second ghost layer CSR adjacencies
```

Comment:
So adjacencies for each ghost layer are stored separately?  How is this array indexed?  How are vertices that are in both the first ghost layer and second ghost layer stored?